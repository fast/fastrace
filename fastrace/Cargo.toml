[package]
name = "fastrace"
version = "0.7.7"

categories = ["development-tools::debugging"]
description = "A high-performance timeline tracing library for Rust"
documentation = "https://docs.rs/fastrace"
keywords = ["tracing", "span", "datadog", "jaeger", "opentelemetry"]
readme = "../README.md"

edition.workspace = true
license.workspace = true
repository.workspace = true
rust-version.workspace = true

[features]
enable = []
serde = ["dep:serde"]

[dependencies]
fastant = "0.1"
fastrace-macro = { workspace = true }
parking_lot = "0.12"
pin-project = { version = "1.1.8" }
rand = "0.9"
rtrb = "0.3"
serde = { version = "1", optional = true }

[dev-dependencies]
async-trait = "0.1"
criterion = { version = "0.5", features = ["html_reports"] }
crossbeam = "0.8"
fastrace = { workspace = true, features = ["enable"] }
fastrace-datadog = { workspace = true }
fastrace-jaeger = { workspace = true }
fastrace-opentelemetry = { workspace = true }
flume = "0.11"
futures-timer = "3"
log = { workspace = true }
logcall = "0.1"
logforth = { version = "0.19.2", features = ["fastrace"] }
mockall = "0.13"
opentelemetry = { version = "0.28", features = ["trace"] }
opentelemetry-otlp = { version = "0.28", features = ["trace", "grpc-tonic"] }
opentelemetry_sdk = { version = "0.28", features = ["trace"] }
pollster = { version = "0.4.0" }
rand = "0.9"
rustracing = "0.6"
serial_test = "3.1"
test-harness = "0.3"
tokio = { version = "1.38", features = [
  "rt",
  "time",
  "macros",
  "rt-multi-thread",
] }
tracing = "0.1"
tracing-opentelemetry = "0.28.0"
tracing-subscriber = "0.3"

[[bench]]
harness = false
name = "trace"

[[bench]]
harness = false
name = "compare"

[[bench]]
harness = false
name = "spsc"

[[bench]]
harness = false
name = "object_pool"
