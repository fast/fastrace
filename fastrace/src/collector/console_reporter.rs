// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.

use crate::collector::SpanRecord;
use crate::collector::global_collector::Reporter;

/// A console reporter that prints span records to the stderr.
pub struct ConsoleReporter;

impl Reporter for ConsoleReporter {
    fn report(&mut self, spans: Vec<SpanRecord>) {
        for span in spans {
            eprintln!("{:#?}", span);
        }
    }
}
