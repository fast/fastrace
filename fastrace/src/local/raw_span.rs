// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.

use std::borrow::Cow;

use fastant::Instant;

use crate::collector::SpanId;
use crate::util::Properties;

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum RawKind {
    Span,
    Event,
    Properties,
}

#[derive(Debug)]
pub struct RawSpan {
    pub id: SpanId,
    pub parent_id: SpanId,
    pub begin_instant: Instant,
    pub name: Cow<'static, str>,
    pub properties: Option<Properties>,
    pub raw_kind: RawKind,

    // Will write this field at post processing
    pub end_instant: Instant,
}

impl RawSpan {
    #[inline]
    pub(crate) fn begin_with(
        id: SpanId,
        parent_id: SpanId,
        begin_instant: Instant,
        name: impl Into<Cow<'static, str>>,
        raw_kind: RawKind,
    ) -> Self {
        RawSpan {
            id,
            parent_id,
            begin_instant,
            name: name.into(),
            properties: None,
            raw_kind,
            end_instant: Instant::ZERO,
        }
    }

    #[inline]
    pub(crate) fn end_with(&mut self, end_instant: Instant) {
        self.end_instant = end_instant;
    }
}

impl Clone for RawSpan {
    fn clone(&self) -> Self {
        let properties = self.properties.as_ref().map(|properties| {
            let mut new_properties = Properties::default();
            new_properties.extend(properties.iter().cloned());
            new_properties
        });

        RawSpan {
            id: self.id,
            parent_id: self.parent_id,
            begin_instant: self.begin_instant,
            name: self.name.clone(),
            properties,
            raw_kind: self.raw_kind,
            end_instant: self.end_instant,
        }
    }
}
