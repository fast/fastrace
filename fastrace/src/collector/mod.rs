// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.

//! Collector and the collected spans.

#![cfg_attr(test, allow(dead_code))]

pub(crate) mod command;
mod console_reporter;
pub(crate) mod global_collector;
pub(crate) mod id;
mod test_reporter;

use std::borrow::Cow;
use std::sync::Arc;
use std::time::Duration;

pub use console_reporter::ConsoleReporter;
#[cfg(not(test))]
pub(crate) use global_collector::GlobalCollect;
#[cfg(test)]
pub(crate) use global_collector::MockGlobalCollect;
pub use global_collector::Reporter;
pub use id::SpanContext;
pub use id::SpanId;
pub use id::TraceId;
#[doc(hidden)]
pub use test_reporter::TestReporter;

use crate::local::local_collector::LocalSpansInner;
use crate::local::raw_span::RawSpan;

#[cfg(test)]
pub(crate) type GlobalCollect = Arc<MockGlobalCollect>;

#[doc(hidden)]
#[derive(Debug)]
pub enum SpanSet {
    Span(RawSpan),
    LocalSpansInner(LocalSpansInner),
    SharedLocalSpans(Arc<LocalSpansInner>),
}

/// A record of a span that includes all the information about the span,
/// such as its identifiers, timing information, name, and associated properties.
#[derive(Clone, Debug, Default, PartialEq)]
pub struct SpanRecord {
    pub trace_id: TraceId,
    pub span_id: SpanId,
    pub parent_id: SpanId,
    pub begin_time_unix_ns: u64,
    pub duration_ns: u64,
    pub name: Cow<'static, str>,
    pub properties: Vec<(Cow<'static, str>, Cow<'static, str>)>,
    pub events: Vec<EventRecord>,
}

/// A record of an event that occurred during the execution of a span.
#[derive(Clone, Debug, Default, PartialEq)]
pub struct EventRecord {
    pub name: Cow<'static, str>,
    pub timestamp_unix_ns: u64,
    pub properties: Vec<(Cow<'static, str>, Cow<'static, str>)>,
}

#[doc(hidden)]
#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub struct CollectTokenItem {
    pub trace_id: TraceId,
    pub parent_id: SpanId,
    pub collect_id: usize,
    pub is_root: bool,
    pub is_sampled: bool,
}

/// Configuration of the behavior of the global collector.
#[must_use]
#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct Config {
    pub(crate) max_spans_per_trace: Option<usize>,
    pub(crate) report_interval: Duration,
    pub(crate) report_before_root_finish: bool,
}

impl Config {
    /// Sets a soft limit for the total number of spans and events in a trace, typically
    /// used to prevent out-of-memory issues.
    ///
    /// The default value is `None`.
    ///
    /// # Note
    ///
    /// The root span will always be collected, so the actual number of collected spans
    /// may exceed the specified limit.
    ///
    /// # Examples
    ///
    /// ```
    /// use fastrace::collector::Config;
    ///
    /// let config = Config::default().max_spans_per_trace(Some(100));
    /// fastrace::set_reporter(fastrace::collector::ConsoleReporter, config);
    /// ```
    pub fn max_spans_per_trace(self, max_spans_per_trace: Option<usize>) -> Self {
        Self {
            max_spans_per_trace,
            ..self
        }
    }

    /// Sets the time duration between two reports. The reporter will be invoked when the specified
    /// duration elapses, even if no spans have been collected. This allows for batching in the
    /// reporter.
    ///
    /// In some scenarios, particularly under high load, you may notice spans being lost. This is
    /// likely due to the channel being full during the reporting interval. To mitigate this issue,
    /// consider reducing the report interval, potentially down to zero, to prevent losing spans.
    ///
    /// The default value is 10 milliseconds.
    ///
    /// # Examples
    ///
    /// ```
    /// use fastrace::collector::Config;
    ///
    /// let config = Config::default().report_interval(std::time::Duration::from_millis(100));
    /// fastrace::set_reporter(fastrace::collector::ConsoleReporter, config);
    /// ```
    pub fn report_interval(self, report_interval: Duration) -> Self {
        Self {
            report_interval,
            ..self
        }
    }

    /// Configures whether to report spans before the root span finishes.
    ///
    /// If set to `true`, some spans may be reported before they are canceled, making it
    /// difficult to cancel all spans in a trace.
    ///
    /// The default value is `false`.
    ///
    /// # Examples
    ///
    /// ```
    /// use fastrace::collector::Config;
    ///
    /// let config = Config::default().report_before_root_finish(true);
    /// fastrace::set_reporter(fastrace::collector::ConsoleReporter, config);
    /// ```
    pub fn report_before_root_finish(self, report_before_root_finish: bool) -> Self {
        Self {
            report_before_root_finish,
            ..self
        }
    }
}

impl Default for Config {
    fn default() -> Self {
        Self {
            max_spans_per_trace: None,
            report_interval: Duration::from_millis(10),
            report_before_root_finish: false,
        }
    }
}

#[cfg(test)]
mod tests {

    use super::*;

    #[test]
    fn w3c_traceparent() {
        let span_context = SpanContext::decode_w3c_traceparent(
            "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01",
        )
        .unwrap();
        assert_eq!(
            span_context.trace_id,
            TraceId(0x0af7651916cd43dd8448eb211c80319c)
        );
        assert_eq!(span_context.span_id, SpanId(0xb7ad6b7169203331));

        assert_eq!(
            span_context.encode_w3c_traceparent(),
            "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01"
        );
        assert_eq!(
            span_context.sampled(false).encode_w3c_traceparent(),
            "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-00"
        );

        assert!(
            !SpanContext::decode_w3c_traceparent(
                "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-00",
            )
            .unwrap()
            .sampled
        );
        assert!(
            SpanContext::decode_w3c_traceparent(
                "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01",
            )
            .unwrap()
            .sampled
        );
        assert!(
            !SpanContext::decode_w3c_traceparent(
                "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-10",
            )
            .unwrap()
            .sampled
        );
    }
}
