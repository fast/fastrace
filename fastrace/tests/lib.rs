// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.

use std::time::Duration;

use fastrace::collector::Config;
use fastrace::collector::ConsoleReporter;
use fastrace::collector::TestReporter;
use fastrace::local::LocalCollector;
use fastrace::prelude::*;
use fastrace::util::tree::tree_str_from_span_records;
use serial_test::serial;
use tokio::runtime::Builder;

fn four_spans() {
    {
        // wide
        for i in 0..2 {
            let _span = LocalSpan::enter_with_local_parent(format!("iter-span-{i}"))
                .with_property(|| ("tmp_property", "tmp_value"));
        }
    }

    {
        #[trace(name = "rec-span")]
        fn rec(mut i: u32) {
            i -= 1;

            if i > 0 {
                rec(i);
            }
        }

        // deep
        rec(2);
    }
}

#[test]
#[serial]
fn single_thread_single_span() {
    let (reporter, collected_spans) = TestReporter::new();
    fastrace::set_reporter(reporter, Config::default());

    {
        let root = Span::root("root", SpanContext::random());
        let _g = root.set_local_parent();

        four_spans();
    };

    fastrace::flush();

    let graph = tree_str_from_span_records(collected_spans.lock().clone());
    insta::assert_snapshot!(graph, @r###"
    root []
        iter-span-0 [("tmp_property", "tmp_value")]
        iter-span-1 [("tmp_property", "tmp_value")]
        rec-span []
            rec-span []
    "###);
}

#[test]
#[serial]
fn single_thread_multiple_spans() {
    let (reporter, collected_spans) = TestReporter::new();
    fastrace::set_reporter(reporter, Config::default());

    {
        let root1 = Span::root("root1", SpanContext::new(TraceId(12), SpanId::default()));
        let root2 = Span::root("root2", SpanContext::new(TraceId(13), SpanId::default()));
        let root3 = Span::root("root3", SpanContext::new(TraceId(14), SpanId::default()));

        let local_collector = LocalCollector::start();

        four_spans();

        let local_spans = local_collector.collect();

        root1.push_child_spans(local_spans.clone());
        root2.push_child_spans(local_spans.clone());
        root3.push_child_spans(local_spans);
    }

    fastrace::flush();

    let graph1 = tree_str_from_span_records(
        collected_spans
            .lock()
            .iter()
            .filter(|s| s.trace_id == TraceId(12))
            .cloned()
            .collect(),
    );
    insta::assert_snapshot!(graph1, @r###"
    root1 []
        iter-span-0 [("tmp_property", "tmp_value")]
        iter-span-1 [("tmp_property", "tmp_value")]
        rec-span []
            rec-span []
    "###);

    let graph2 = tree_str_from_span_records(
        collected_spans
            .lock()
            .iter()
            .filter(|s| s.trace_id == TraceId(13))
            .cloned()
            .collect(),
    );
    insta::assert_snapshot!(graph2, @r###"
    root2 []
        iter-span-0 [("tmp_property", "tmp_value")]
        iter-span-1 [("tmp_property", "tmp_value")]
        rec-span []
            rec-span []
    "###);

    let graph3 = tree_str_from_span_records(
        collected_spans
            .lock()
            .iter()
            .filter(|s| s.trace_id == TraceId(14))
            .cloned()
            .collect(),
    );
    insta::assert_snapshot!(graph3, @r###"
    root3 []
        iter-span-0 [("tmp_property", "tmp_value")]
        iter-span-1 [("tmp_property", "tmp_value")]
        rec-span []
            rec-span []
    "###);
}

#[test]
#[serial]
fn multiple_threads_single_span() {
    let (reporter, collected_spans) = TestReporter::new();
    fastrace::set_reporter(reporter, Config::default());

    crossbeam::scope(|scope| {
        let root = Span::root("root", SpanContext::random());
        let _g = root.set_local_parent();

        let mut handles = vec![];

        for _ in 0..4 {
            let child_span = Span::enter_with_local_parent("cross-thread");
            let h = scope.spawn(move |_| {
                let _g = child_span.set_local_parent();
                four_spans();
            });
            handles.push(h);
        }

        four_spans();

        handles.into_iter().for_each(|h| h.join().unwrap());
    })
    .unwrap();

    fastrace::flush();

    let graph = tree_str_from_span_records(collected_spans.lock().clone());
    insta::assert_snapshot!(graph, @r###"
    root []
        cross-thread []
            iter-span-0 [("tmp_property", "tmp_value")]
            iter-span-1 [("tmp_property", "tmp_value")]
            rec-span []
                rec-span []
        cross-thread []
            iter-span-0 [("tmp_property", "tmp_value")]
            iter-span-1 [("tmp_property", "tmp_value")]
            rec-span []
                rec-span []
        cross-thread []
            iter-span-0 [("tmp_property", "tmp_value")]
            iter-span-1 [("tmp_property", "tmp_value")]
            rec-span []
                rec-span []
        cross-thread []
            iter-span-0 [("tmp_property", "tmp_value")]
            iter-span-1 [("tmp_property", "tmp_value")]
            rec-span []
                rec-span []
        iter-span-0 [("tmp_property", "tmp_value")]
        iter-span-1 [("tmp_property", "tmp_value")]
        rec-span []
            rec-span []
    "###);
}

#[test]
#[serial]
fn multiple_threads_multiple_spans() {
    let (reporter, collected_spans) = TestReporter::new();
    fastrace::set_reporter(reporter, Config::default());

    crossbeam::scope(|scope| {
        let root1 = Span::root("root1", SpanContext::new(TraceId(12), SpanId::default()));
        let root2 = Span::root("root2", SpanContext::new(TraceId(13), SpanId::default()));
        let local_collector = LocalCollector::start();

        let mut handles = vec![];

        for _ in 0..4 {
            let merged = Span::enter_with_parents("merged", vec![&root1, &root2]);
            let _g = merged.set_local_parent();
            let _local = LocalSpan::enter_with_local_parent("local");
            let h = scope.spawn(move |_| {
                let local_collector = LocalCollector::start();

                four_spans();

                let local_spans = local_collector.collect();
                merged.push_child_spans(local_spans);
            });

            handles.push(h);
        }

        four_spans();

        handles.into_iter().for_each(|h| h.join().unwrap());

        let local_spans = local_collector.collect();
        root1.push_child_spans(local_spans.clone());
        root2.push_child_spans(local_spans);
    })
    .unwrap();

    fastrace::flush();

    let graph1 = tree_str_from_span_records(
        collected_spans
            .lock()
            .iter()
            .filter(|s| s.trace_id == TraceId(12))
            .cloned()
            .collect(),
    );
    insta::assert_snapshot!(graph1, @r###"
    root1 []
        iter-span-0 [("tmp_property", "tmp_value")]
        iter-span-1 [("tmp_property", "tmp_value")]
        merged []
            iter-span-0 [("tmp_property", "tmp_value")]
            iter-span-1 [("tmp_property", "tmp_value")]
            local []
            rec-span []
                rec-span []
        merged []
            iter-span-0 [("tmp_property", "tmp_value")]
            iter-span-1 [("tmp_property", "tmp_value")]
            local []
            rec-span []
                rec-span []
        merged []
            iter-span-0 [("tmp_property", "tmp_value")]
            iter-span-1 [("tmp_property", "tmp_value")]
            local []
            rec-span []
                rec-span []
        merged []
            iter-span-0 [("tmp_property", "tmp_value")]
            iter-span-1 [("tmp_property", "tmp_value")]
            local []
            rec-span []
                rec-span []
        rec-span []
            rec-span []
    "###);

    let graph2 = tree_str_from_span_records(
        collected_spans
            .lock()
            .iter()
            .filter(|s| s.trace_id == TraceId(13))
            .cloned()
            .collect(),
    );
    insta::assert_snapshot!(graph2, @r###"
    root2 []
        iter-span-0 [("tmp_property", "tmp_value")]
        iter-span-1 [("tmp_property", "tmp_value")]
        merged []
            iter-span-0 [("tmp_property", "tmp_value")]
            iter-span-1 [("tmp_property", "tmp_value")]
            local []
            rec-span []
                rec-span []
        merged []
            iter-span-0 [("tmp_property", "tmp_value")]
            iter-span-1 [("tmp_property", "tmp_value")]
            local []
            rec-span []
                rec-span []
        merged []
            iter-span-0 [("tmp_property", "tmp_value")]
            iter-span-1 [("tmp_property", "tmp_value")]
            local []
            rec-span []
                rec-span []
        merged []
            iter-span-0 [("tmp_property", "tmp_value")]
            iter-span-1 [("tmp_property", "tmp_value")]
            local []
            rec-span []
                rec-span []
        rec-span []
            rec-span []
    "###);
}

#[test]
#[serial]
fn multiple_spans_without_local_spans() {
    let (reporter, collected_spans) = TestReporter::new();
    fastrace::set_reporter(reporter, Config::default().cancelable(true));

    {
        let root1 = Span::root("root1", SpanContext::new(TraceId(12), SpanId::default()));
        let root2 = Span::root("root2", SpanContext::new(TraceId(13), SpanId::default()));
        let root3 = Span::root("root3", SpanContext::new(TraceId(14), SpanId::default()));

        let local_collector = LocalCollector::start();

        let local_spans = local_collector.collect();
        root1.push_child_spans(local_spans.clone());
        root2.push_child_spans(local_spans.clone());
        root3.push_child_spans(local_spans);

        root3.cancel();
    }

    fastrace::flush();

    assert_eq!(
        collected_spans
            .lock()
            .iter()
            .filter(|s| s.trace_id == TraceId(12))
            .count(),
        1
    );
    assert_eq!(
        collected_spans
            .lock()
            .iter()
            .filter(|s| s.trace_id == TraceId(13))
            .count(),
        1
    );
    assert_eq!(
        collected_spans
            .lock()
            .iter()
            .filter(|s| s.trace_id == TraceId(14))
            .count(),
        0
    );
}

#[test]
#[serial]
fn test_macro() {
    use async_trait::async_trait;

    #[async_trait]
    trait Foo {
        async fn run(&self, millis: &u64);
    }

    struct Bar;

    #[async_trait]
    impl Foo for Bar {
        #[trace(name = "run")]
        async fn run(&self, millis: &u64) {
            let _g = Span::enter_with_local_parent("run-inner");
            work(millis).await;
            let _g = LocalSpan::enter_with_local_parent("local-span");
        }
    }

    #[trace(short_name = true, enter_on_poll = true)]
    async fn work(millis: &u64) {
        let _g = Span::enter_with_local_parent("work-inner");
        tokio::time::sleep(Duration::from_millis(*millis))
            .enter_on_poll("sleep")
            .await;
    }

    impl Bar {
        #[trace(short_name = true)]
        async fn work2(&self, millis: &u64) {
            let _g = Span::enter_with_local_parent("work-inner");
            tokio::time::sleep(Duration::from_millis(*millis))
                .enter_on_poll("sleep")
                .await;
        }
    }

    #[trace(short_name = true)]
    async fn work3(millis1: &u64, millis2: &u64) {
        let _g = Span::enter_with_local_parent("work-inner");
        tokio::time::sleep(Duration::from_millis(*millis1))
            .enter_on_poll("sleep")
            .await;
        tokio::time::sleep(Duration::from_millis(*millis2))
            .enter_on_poll("sleep")
            .await;
    }

    let (reporter, collected_spans) = TestReporter::new();
    fastrace::set_reporter(reporter, Config::default());

    {
        let root = Span::root("root", SpanContext::random());

        let runtime = Builder::new_multi_thread()
            .worker_threads(4)
            .enable_all()
            .build()
            .unwrap();

        pollster::block_on(
            runtime.spawn(
                async {
                    Bar.run(&100).await;
                    Bar.work2(&100).await;
                    work3(&100, &100).await;
                }
                .in_span(root),
            ),
        )
        .unwrap();
    }

    fastrace::flush();

    let graph = tree_str_from_span_records(collected_spans.lock().clone());
    insta::assert_snapshot!(graph, @r###"
    root []
        run []
            local-span []
            run-inner []
            work []
                sleep []
            work []
                sleep []
                work-inner []
        work2 []
            sleep []
            sleep []
            work-inner []
        work3 []
            sleep []
            sleep []
            sleep []
            sleep []
            work-inner []
    "###);
}

#[test]
#[serial]
fn macro_example() {
    #[trace(short_name = true)]
    fn do_something_short_name(i: u64) {
        std::thread::sleep(Duration::from_millis(i));
    }

    #[trace(short_name = true)]
    async fn do_something_async_short_name(i: u64) {
        futures_timer::Delay::new(Duration::from_millis(i)).await;
    }

    #[trace]
    fn do_something(i: u64) {
        std::thread::sleep(Duration::from_millis(i));
    }

    #[trace]
    async fn do_something_async(i: u64) {
        futures_timer::Delay::new(Duration::from_millis(i)).await;
    }

    let (reporter, collected_spans) = TestReporter::new();
    fastrace::set_reporter(reporter, Config::default());

    {
        let root = Span::root("root", SpanContext::random());
        let _g = root.set_local_parent();
        do_something(100);
        pollster::block_on(do_something_async(100));
        do_something_short_name(100);
        pollster::block_on(do_something_async_short_name(100));
    }

    fastrace::flush();

    let graph = tree_str_from_span_records(collected_spans.lock().clone());
    insta::assert_snapshot!(graph,@r###"
    root []
        do_something_async_short_name []
        do_something_short_name []
        lib::macro_example::{{closure}}::do_something []
        lib::macro_example::{{closure}}::do_something_async::{{closure}} []
    "###);
}

#[test]
#[serial]
fn multiple_local_parent() {
    let (reporter, collected_spans) = TestReporter::new();
    fastrace::set_reporter(reporter, Config::default());

    {
        let root = Span::root("root", SpanContext::random());
        let _g = root.set_local_parent();
        let _g = LocalSpan::enter_with_local_parent("span1");
        let span2 = Span::enter_with_local_parent("span2");
        {
            let _g = span2.set_local_parent();
            let _g = LocalSpan::enter_with_local_parent("span3");
        }
        let _g = LocalSpan::enter_with_local_parent("span4");
    }

    fastrace::flush();

    let graph = tree_str_from_span_records(collected_spans.lock().clone());
    insta::assert_snapshot!(graph, @r###"
    root []
        span1 []
            span2 []
                span3 []
            span4 []
    "###);
}

#[test]
#[serial]
fn early_local_collect() {
    let (reporter, collected_spans) = TestReporter::new();
    fastrace::set_reporter(reporter, Config::default());

    {
        let local_collector = LocalCollector::start();
        let _g1 = LocalSpan::enter_with_local_parent("span1");
        let _g2 = LocalSpan::enter_with_local_parent("span2");
        drop(_g2);
        let local_spans = local_collector.collect();

        let root = Span::root("root", SpanContext::random());
        root.push_child_spans(local_spans);
    }

    fastrace::flush();

    let graph = tree_str_from_span_records(collected_spans.lock().clone());
    insta::assert_snapshot!(graph, @r###"
    root []
        span1 []
            span2 []
    "###);
}

#[test]
#[serial]
fn test_elapsed() {
    fastrace::set_reporter(ConsoleReporter, Config::default());

    {
        let root = Span::root("root", SpanContext::random());

        std::thread::sleep(Duration::from_millis(50));

        assert!(root.elapsed().unwrap() >= Duration::from_millis(50));
    }

    fastrace::flush();
}

#[test]
#[serial]
fn test_property() {
    let (reporter, collected_spans) = TestReporter::new();
    fastrace::set_reporter(reporter, Config::default());

    {
        let root = Span::root("root", SpanContext::random())
            .with_property(|| ("k1", "v1"))
            .with_properties(|| [("k2", "v2"), ("k3", "v3")]);
        root.add_property(|| ("k4", "v4"));
        root.add_properties(|| [("k5", "v5"), ("k6", "v6")]);
        let _g = root.set_local_parent();
        LocalSpan::add_property(|| ("k7", "v7"));
        LocalSpan::add_properties(|| [("k8", "v8"), ("k9", "v9")]);
        let _span = LocalSpan::enter_with_local_parent("span")
            .with_property(|| ("k10", "v10"))
            .with_properties(|| [("k11", "v11"), ("k12", "v12")]);
        LocalSpan::add_property(|| ("k13", "v13"));
        LocalSpan::add_properties(|| [("k14", "v14"), ("k15", "v15")]);
    }

    fastrace::flush();

    let graph = tree_str_from_span_records(collected_spans.lock().clone());
    insta::assert_snapshot!(graph, @r###"
    root [("k1", "v1"), ("k2", "v2"), ("k3", "v3"), ("k4", "v4"), ("k5", "v5"), ("k6", "v6"), ("k7", "v7"), ("k8", "v8"), ("k9", "v9")]
        span [("k10", "v10"), ("k11", "v11"), ("k12", "v12"), ("k13", "v13"), ("k14", "v14"), ("k15", "v15")]
    "###);
}

#[test]
#[serial]
fn test_event() {
    let (reporter, collected_spans) = TestReporter::new();
    fastrace::set_reporter(reporter, Config::default());

    {
        let root = Span::root("root", SpanContext::random());
        root.add_event(
            Event::new("event1 in root")
                .with_property(|| ("k1", "v1"))
                .with_properties(|| [("k2", "v2"), ("k3", "v3")]),
        );
        let _g = root.set_local_parent();
        LocalSpan::add_event(
            Event::new("event2 in root")
                .with_property(|| ("k4", "v4"))
                .with_properties(|| [("k5", "v5"), ("k6", "v6")]),
        );
        let _span = LocalSpan::enter_with_local_parent("span");
        LocalSpan::add_event(
            Event::new("event3 in span")
                .with_property(|| ("k7", "v7"))
                .with_properties(|| [("k8", "v8"), ("k9", "v9")]),
        );
    }

    fastrace::flush();

    let graph = tree_str_from_span_records(collected_spans.lock().clone());
    insta::assert_snapshot!(graph, @r###"
    root [] [("event1 in root", [("k1", "v1"), ("k2", "v2"), ("k3", "v3")]), ("event2 in root", [("k4", "v4"), ("k5", "v5"), ("k6", "v6")])]
        span [] [("event3 in span", [("k7", "v7"), ("k8", "v8"), ("k9", "v9")])]
    "###);
}

#[test]
#[serial]
fn test_macro_properties() {
    #[allow(clippy::drop_non_drop)]
    #[trace(short_name = true, properties = { "k1": "v1", "a": "argument a is {a:?}", "b": "{b:?}", "escaped1": "{c:?}{{}}", "escaped2": "{{ \"a\": \"b\"}}" })]
    fn foo(a: i64, b: &Bar, c: Bar) {
        drop(c);
    }

    #[allow(clippy::drop_non_drop)]
    #[trace(short_name = true, properties = { "k1": "v1", "a": "argument a is {a:?}", "b": "{b:?}", "escaped1": "{c:?}{{}}", "escaped2": "{{ \"a\": \"b\"}}" })]
    async fn foo_async(a: i64, b: &Bar, c: Bar) {
        drop(c);
    }

    #[trace(short_name = true, properties = {})]
    fn bar() {}

    #[trace(short_name = true, properties = {})]
    async fn bar_async() {}

    #[derive(Debug)]
    struct Bar;

    let (reporter, collected_spans) = TestReporter::new();
    fastrace::set_reporter(reporter, Config::default());

    {
        let root = Span::root("root", SpanContext::random());
        let _g = root.set_local_parent();
        foo(1, &Bar, Bar);
        bar();

        let runtime = Builder::new_multi_thread()
            .worker_threads(4)
            .enable_all()
            .build()
            .unwrap();

        pollster::block_on(
            runtime.spawn(
                async {
                    foo_async(1, &Bar, Bar).await;
                    bar_async().await;
                }
                .in_span(root),
            ),
        )
        .unwrap();
    }

    fastrace::flush();

    let graph = tree_str_from_span_records(collected_spans.lock().clone());
    insta::assert_snapshot!(graph, @r###"
    root []
        bar []
        bar_async []
        foo [("k1", "v1"), ("a", "argument a is 1"), ("b", "Bar"), ("escaped1", "Bar{}"), ("escaped2", "{ \"a\": \"b\"}")]
        foo_async [("k1", "v1"), ("a", "argument a is 1"), ("b", "Bar"), ("escaped1", "Bar{}"), ("escaped2", "{ \"a\": \"b\"}")]
    "###);
}

#[test]
#[serial]
fn test_not_sampled() {
    let (reporter, collected_spans) = TestReporter::new();
    fastrace::set_reporter(reporter, Config::default());
    {
        let root = Span::root("root", SpanContext::random().sampled(true));
        let _g = root.set_local_parent();
        let _span = LocalSpan::enter_with_local_parent("span");
    }
    fastrace::flush();

    let graph = tree_str_from_span_records(collected_spans.lock().clone());
    insta::assert_snapshot!(graph,@r###"
    root []
        span []
    "###);

    let (reporter, collected_spans) = TestReporter::new();
    fastrace::set_reporter(reporter, Config::default());
    {
        let root = Span::root("root", SpanContext::random().sampled(false));
        let _g = root.set_local_parent();
        let _span = LocalSpan::enter_with_local_parent("span");
    }
    fastrace::flush();
    assert!(collected_spans.lock().is_empty());
}
