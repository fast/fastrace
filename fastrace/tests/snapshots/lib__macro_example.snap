---
source: fastrace/tests/lib.rs
expression: graph
---
root []
    do_something_async_short_name []
    do_something_short_name []
    lib::macro_example::{{closure}}::do_something []
    lib::macro_example::{{closure}}::do_something_async::{{closure}} []
